### ISSUE TITLE
Plans for v3.5.3 release

### ISSUE THREAD

**Issue by serathius** [etcd-io/etcd#13894]
Same as previous release https://github.com/etcd-io/etcd/issues/13518. let's agree what work should be included in release and track the work. The list presented below will be evolving as community decides to cut/include some fixes.

Issues/PRs that should be included for v3.5.3:

* [x] Qualify the release after all cherry-picks are done
  * [x] Run scripts used for reproduction of data inconsistencies found in v3.5. https://github.com/etcd-io/etcd/issues/13766 https://github.com/etcd-io/etcd/issues/13514 (@serathius)
  * [x] Run manual heavier functional tests (with https://github.com/etcd-io/etcd/pull/13926, https://github.com/etcd-io/etcd/pull/13924 and https://github.com/etcd-io/etcd/pull/13927) (@serathius) 
  * [x] Run [jepsen tests ](https://github.com/jepsen-io/etcd)(@endocrimes) 
* [x] https://github.com/etcd-io/etcd/issues/13922
  * [x] https://github.com/etcd-io/etcd/pull/13930
  * [x] https://github.com/etcd-io/etcd/pull/13933
* [x] https://github.com/etcd-io/etcd/issues/13766 - data inconsistency issue caused by SIGKILL
  * [x] https://github.com/etcd-io/etcd/pull/13887
  * [x] https://github.com/etcd-io/etcd/pull/13854
  * [x] https://github.com/etcd-io/etcd/pull/13908
* [x] https://github.com/etcd-io/etcd/issues/13514 - data inconsistency issue caused by snapshot restore
  * [x] https://github.com/etcd-io/etcd/pull/13903
  * [x] https://github.com/etcd-io/etcd/pull/13904
* [x] https://github.com/etcd-io/etcd/issues/13853 
  * [x] https://github.com/etcd-io/etcd/pull/13837
  * [x] https://github.com/etcd-io/etcd/pull/13917
* [x] https://github.com/etcd-io/etcd/pull/13669
  * [x] https://github.com/etcd-io/etcd/pull/13910
* [x] https://github.com/etcd-io/etcd/pull/13895
* [x] https://github.com/etcd-io/etcd/pull/13866
* [x] https://github.com/etcd-io/etcd/pull/13832
* [x] https://github.com/etcd-io/etcd/pull/13727
* [x] https://github.com/etcd-io/etcd/pull/13714
* [x] https://github.com/etcd-io/etcd/pull/13706
* [x] https://github.com/etcd-io/etcd/issues/13675 
  * [x] https://github.com/etcd-io/etcd/pull/13690
  * [x] https://github.com/etcd-io/etcd/pull/13932

Issues/PRs not required for v3.5.3:

* https://github.com/etcd-io/etcd/pull/13861
* https://github.com/etcd-io/etcd/issues/13883
* https://github.com/etcd-io/etcd/issues/13850
* https://github.com/etcd-io/etcd/issues/13757
* https://github.com/etcd-io/etcd/pull/13748
* https://github.com/etcd-io/etcd/issues/13666
* https://github.com/etcd-io/etcd/issues/13692
* https://github.com/etcd-io/etcd/issues/13848
* https://github.com/etcd-io/etcd/issues/13879
* https://github.com/etcd-io/etcd/pull/13862
* https://github.com/etcd-io/etcd/issues/13846

Please feel free to suggest new additional backports/issues worth investigating so we can discuss if they should be included.

**Comment by serathius** [comment]
cc @ptabor @ahrtr @spzala

**Comment by mrueg** [comment]
@serathius not sure if you want to add https://github.com/etcd-io/etcd/pull/13895 as well.

**Comment by ahrtr** [comment]
I think we need [13838](https://github.com/etcd-io/etcd/pull/13838) as well, because it will make the reproduction of the data inconsistency issue easier.

**Comment by serathius** [comment]
Hey @ahrtr, It would be great to have automatic way to prevent data inconsistencies in the future, however this is not required immediate release. https://github.com/etcd-io/etcd/pull/13838 includes changes that might negatively impact overall stability of functional tests making it hard to merge. I don't plan to make this code mergeable, to avoid investing to much time into functional testing as we might want to change our overall approach to correctness testing after the data inconsistency issue.

The data inconsistency issue should be fairly tested with https://github.com/etcd-io/etcd/pull/13838, but I don't see merging this change as a blocking for v3.5.3.

**Comment by riyazcom** [comment]
@serathius  is it possible to include [#13460 ](https://github.com/etcd-io/etcd/pull/13460) "Support multiple values for allowed client and peer TLS identities"

**Comment by serathius** [comment]
@riyazcom Even though not supporting multiple client & peer TLS identities is a limitation of etcd, it is not a bug and introducing it would be basically be an additional feature. Based on that I don't think we should back port.

**Comment by jayunit100** [comment]
what is the manual test plan for reproducing this ? is there a k8s centric test we can run ?  I was doing a combination of running e2es while rebooting nodes, and didnt see this occur.  I think it requires unrestricted API QoS , constrained etcd memory, and a few other tricky knobs.

**Comment by serathius** [comment]
Please read the original issue https://github.com/etcd-io/etcd/issues/13766#issuecomment-1078897588, it includes description of reproduction and link to PR with code. 

Let's continue the discussion there ok?

**Comment by spzala** [comment]
@serathius https://github.com/etcd-io/etcd/issues/13692 doesn't seems required in 3.5.3. I am adding triage comments there.

**Comment by ahrtr** [comment]
[13666](https://github.com/etcd-io/etcd/issues/13666) is just a question instead of an issue. I already provided answer and closed it. Definitely it isn't required in 3.5.3.

**Comment by ahrtr** [comment]
Please also remove [13690](https://github.com/etcd-io/etcd/pull/13690) from the list per the discussion in the PR.

**Comment by ahrtr** [comment]
Just closed 13848 with comment [13848#issuecomment-1093449532](https://github.com/etcd-io/etcd/issues/13848#issuecomment-1093449532).

**Comment by yuyangbj** [comment]
Hi ETCD maitainers,

As the community has alert the customers not use the ETCD 3.5.[0-2] in production env due to the issue data inconsistency, but is there any plan to speed up the 3.5.3 release considering some products has already blocked by the issue?

**Comment by serathius** [comment]
@yuyangbj Thanks for your interest in v3.5.3 release. Apart of just "fixing" the issue, we also need to make sure we do proper diligence to test and qualify the release. We would really not want to make a mistake here and be forced to send another alert. Unfortunately this takes time and with only 3-4 people working on the release it will take us some time.

Data inconsistency issue was non-trivial and it took us longer time to figure out a proper fix. Fortunately we finally managed to merge the fix last Friday, so we can start qualifying the release and decide to skip some minor issues proposed here. 

Best thing I can recommend for products blocked on the release is stop waiting passively and get involved into the release proces. We need more people to test, document and qualify the release.

**Comment by dims** [comment]
@serathius Apologies!!

**Comment by serathius** [comment]
New issue that might impact v3.5.3 release. Please someone take a look https://github.com/etcd-io/etcd/issues/13922

**Comment by endocrimes** [comment]
https://github.com/etcd-io/etcd/issues/13846 shouldn't be a critical release blocker - but here's a PR https://github.com/etcd-io/etcd/pull/13923

**Comment by serathius** [comment]
Thanks @endocrimes, agree it's not a critical release blocker, however for now let's assume it will go in, as qualification will take some time.

**Comment by endocrimes** [comment]
@serathius 👍  - what's necessary for qualification? (not sure if there's a test process that gets followed or if setting up some soak/scale tests out-of-tree would be helpful for example)

**Comment by serathius** [comment]
That's great question, apparently v3.5 release haven't followed the same qualification as previous releases (https://github.com/etcd-io/etcd/issues/13839#issuecomment-1079823211). With maintainer rotation we lost knowlegde about processes previously executed manually. 

I would be great to get idea what kind of tests were done before. @xiang90 @jpbetz any recommendation about the testing?

**Comment by serathius** [comment]
In addition to the qualification mentioned before, would be good to confirm that two data inconsistency issues found were fixed. Instructions for their reproduction:
* https://github.com/etcd-io/etcd/pull/13838
* https://github.com/etcd-io/etcd/issues/13514#issuecomment-1085784591

**Comment by serathius** [comment]
> @serathius Have we in the past have some method to try a k8s (presubmit?) testing with a branch of etcd? (curious, if so, then we could try that?)

We run v3.5.0 in K8s presubmits for long time (we still are), however they didn't find any issues we fixed for v3.5.3 so I expect they are not good qualification method.

**Comment by yuyangbj** [comment]
> @yuyangbj Thanks for your interest in v3.5.3 release. Apart of just "fixing" the issue, we also need to make sure we do proper diligence to test and qualify the release. We would really not want to make a mistake here and be forced to send another alert. Unfortunately this takes time and with only 3-4 people working on the release it will take us some time.
> 
> Data inconsistency issue was non-trivial and it took us longer time to figure out a proper fix. Fortunately we finally managed to merge the fix last Friday, so we can start qualifying the release and decide to skip some minor issues proposed here.
> 
> Best thing I can recommend for products blocked on the release is stop waiting passively and get involved into the release proces. We need more people to test, document and qualify the release.

Thanks Marek @serathius for the prompt response. And I understand everything needs a complete tests. I do not want to push the release, so I am sorry if it makes a misunderstanding like it. And our downstream team has actively involved to reproduce the data inconsistency issue, will update to here if we have test results.

**Comment by ahrtr** [comment]
> New issue that might impact v3.5.3 release. Please someone take a look #13922

Just resolved in [pull/13930](https://github.com/etcd-io/etcd/pull/13930), it should be legacy issue (or an old regression) instead of a regression.

**Comment by serathius** [comment]
With last data inconsistency issue fixed we can focus on qualification and cutting the scope for the release. Moving https://github.com/etcd-io/etcd/pull/13862 and https://github.com/etcd-io/etcd/issues/13846 as they are not critical. Authors can continue to work on them, however we should prioritize the release.

**Comment by serathius** [comment]
@shalinmangar in https://github.com/etcd-io/etcd/issues/13839#issuecomment-1085949492 you mentioned that you have experience with jepsen. It would be great if you can help @endocrimes as we cannot reproduce the data inconsistencies from v3.5.2 there.

**Comment by serathius** [comment]
Qualification work is almost finished, I would like to prepare to start the release process. 

First let's collect acts from people involved: @ptabor @spzala @ahrtr @endocrimes 
Please let me know if there is anything else that should be done before the release.

**Comment by endocrimes** [comment]
So current state of jepsen testing:
- I have the existing jepsen test suite [jepsen-io/etcd](https://github.com/jepsen-io/etcd) bumped to the latest parts of various etcd components
- I've run the full suite against etcd 3.5.2 in various scale scenarios trying to trigger a corruption case (and failed to do so on my 32t/128GB ram workstation without running out of memory first)
- I'm now setting up an env to run the 3.5.3 release branch through regression scenarios, should be done in an hour or so.

Future plan:
- I've started working on cases that would potentially be able to catch future cases, but probably won't be done before 3.5.3 is ready. Going to flesh this out so we can have regular testing of main somewhere.
- Going to think more deeply about how we can do scale/soak testing for future releases (aside from just jepsen). [need to file an issue]

**Comment by ptabor** [comment]
That list looks complete to me. Thank you @serathius and @ahrtr. 

I think that these items are below the bar (but calling them explicitly): 
  - submit  & backport the data-inconsistency repro test 
  - more defence in depth (e.g. verification that consistency index does not decrease)

@serathius: Out of curiosity: Have you been running the repro test for several hours with patched 3.5.2 or just the O(10)min test ?

**Comment by endocrimes** [comment]
Jepsen Watch test failures _seem_ to be a jepsen issue, but I haven't run into those issues in any of the previous version runs, so I'm... curious.

Moving on to lock tests for now.

**Comment by endocrimes** [comment]
Locks (leases) seem to be consistently flaky - The lease ends up closing early, and thus the test fails. Gonna investigate further after dinner.

**Comment by endocrimes** [comment]
Gonna look into this with fresh eyes tomorrow - getting a headache.

**Comment by dims** [comment]
Sooo close!!

**Comment by ahrtr** [comment]
I think it's worth to mention @serathius 's PR [pull/13927](https://github.com/etcd-io/etcd/pull/13927). Ran the load test against mounted RAM filesystem might be an easier way to reproduce the data inconsistent issue. I will give it a try.

> Locks (leases) seem to be consistently flaky - The lease ends up closing early, and thus the test fails. Gonna investigate further after dinner.

Please note that I recently fixed a lease related issue, see [pull/13932](https://github.com/etcd-io/etcd/pull/13932), which isn't included in 3.5.2. Do you still see flaky on release-3.5 (3.5.3)? If yes, please file an issue, and let's discuss it there. 


> submit & backport the data-inconsistency repro test
@serathius already submitted a [PR](https://github.com/etcd-io/etcd/pull/13924) for this, and I will take a closer review today & tomorrow.


> more defence in depth (e.g. verification that consistency index does not decrease)
Yes, it's a good point. The consistent_index should never decrease based on current design/implementation. But the term might be. I am not totally convinced that the PR [pull/13903](https://github.com/etcd-io/etcd/pull/13903) fixes the issue [issues/13514](https://github.com/etcd-io/etcd/issues/13514), because raft should reject a message with old term.

**Comment by spzala** [comment]
> Qualification work is almost finished, I would like to prepare to start the release process.
> 
> First let's collect acts from people involved: @ptabor @spzala @ahrtr @endocrimes Please let me know if there is anything else that should be done before the release.

@serathius awesome, thanks so much! It looks good to me. If we are running more tests, that's great, but otherwise, from the positive feedback on the data inconsistency fix and the tests you and @ahrtr ran, +1 for release work. We can iterate over the release with any new findings, post-release feedback and test improvements under discussion.

**Comment by serathius** [comment]
> @serathius: Out of curiosity: Have you been running the repro test for several hours with patched 3.5.2 or just the O(10)min test ?

I run:
* Reproduction script https://github.com/etcd-io/etcd/issues/13766 for 1 minute 10 times, reproduces 5/10 times on `v3.5.2` and 0/10 times on current `release-3.5` branch
* Reproduction script for https://github.com/etcd-io/etcd/issues/13514 5 times, reproduces 3/5 times on `v3.5.2` and 0/5 times on current `release-3.5` branch.
* Full functional tests with addition of (https://github.com/etcd-io/etcd/pull/13926, https://github.com/etcd-io/etcd/pull/13924) and ETCD_VERIFY enabled (not enabled by default?) on both ssd disk and ramfs on `release-3.5` branch. Tests failed however not due to any data corruption. 
* Full integration tests on both ssd disk and ramfs on `release-3.5` branch. Tests passed.

**Comment by serathius** [comment]
> Yes, it's a good point. The consistent_index should never decrease based on current design/implementation. But the term might be. I am not totally convinced that the PR [pull/13903](https://github.com/etcd-io/etcd/pull/13903) fixes the issue [issues/13514](https://github.com/etcd-io/etcd/issues/13514), because raft should reject a message with old term.

You are right that raft should never allow a message with an old term, however https://github.com/etcd-io/etcd/issues/13514 covers not officially supported downgrade flow, that went around downgrade check, and showed us that term in db cannot be trusted in v3.5. With https://github.com/etcd-io/etcd/pull/13903 goal was not to fix downgrades, but to remove incorrect `onlyGrow` check that turned out to be incorrectly introduced during rafactor.

**Comment by ahrtr** [comment]
We might need to get this issue [13937](https://github.com/etcd-io/etcd/issues/13937) sorted out before releasing 3.5.3.

**Comment by serathius** [comment]
@ahrtr Is this a new problem in v3.5? Based on the original issue it seems to be present in v3.3. For v3.5.3 our goal should be addressing new issues that were introduced in v3.5 to ensure there are not regressions when upgrading from v3.4 to v3.5. 

Fact that there is a long standing issue is worrying, however it should not block v3.5.3 release.

**Comment by shalinmangar** [comment]
> @shalinmangar in [#13839 (comment)](https://github.com/etcd-io/etcd/issues/13839#issuecomment-1085949492) you mentioned that you have experience with jepsen. It would be great if you can help @endocrimes as we cannot reproduce the data inconsistencies from v3.5.2 there.

Happy to help. I see @endocrimes has already made great progress. Is there an issue for Jepsen testing? or should we communicate over the Etcd-dev mailing list?

**Comment by endocrimes** [comment]
@shalinmangar https://github.com/etcd-io/etcd/issues/13939

**Comment by dims** [comment]
w00t! sounds good @serathius

**Comment by serathius** [comment]
ok, after 4 hours, with breaks for restoring sanity and eating, I managed to successfully run release script.

**Comment by ahrtr** [comment]
> @ahrtr Is this a new problem in v3.5? Based on the original issue it seems to be present in v3.3. For v3.5.3 our goal should be addressing new issues that were introduced in v3.5 to ensure there are not regressions when upgrading from v3.4 to v3.5.
> 
> Fact that there is a long standing issue is worrying, however it should not block v3.5.3 release.

My immediate feeling is it should be legacy issue instead of a regression. Will take a closer look today.  I don't think it's a blocker for 3.5.3 either.

**Comment by ahrtr** [comment]
> @ahrtr Is this a new problem in v3.5? Based on the original issue it seems to be present in v3.3. For v3.5.3 our goal should be addressing new issues that were introduced in v3.5 to ensure there are not regressions when upgrading from v3.4 to v3.5.
> 
> Fact that there is a long standing issue is worrying, however it should not block v3.5.3 release.

This turns out to be regression. But the good news is it's just a minor issue. Please see the explanation in [pull/13942](https://github.com/etcd-io/etcd/pull/13942)
